nvse/nvse/NVSECompiler.cpp

```diff
@@ -0,0 +1,760 @@
+#include "NVSECompiler.h"
+#include "Commands_MiscRef.h"
+#include "Commands_Scripting.h"
+#include "NVSECompilerUtils.h"
+#include "NVSEParser.h"
+
+bool NVSECompiler::compile(Script* script, NVSEScript& ast, std::function<void(std::string)> printFn) {
+	this->script = script;
+	originalScriptName = script->GetEditorID();
+	this->printFn = printFn;
+
+	insideNvseExpr.push(false);
+	ast.accept(this);
+
+	script->SetEditorID(scriptName.c_str());
+
+	script->info.compiled = true;
+	script->info.dataLength = data.size();
+	script->info.numRefs = script->refList.Count();
+	script->info.varCount = script->varList.Count();
+	script->info.unusedVariableCount = script->info.varCount - usedVars.size();
+	script->data = static_cast<uint8_t*>(FormHeap_Allocate(data.size()));
+	memcpy(script->data, data.data(), data.size());
+
+	// Debug print local info
+	printFn("[Locals]\n");
+	for (int i = 0; i < script->varList.Count(); i++) {
+		auto item = script->varList.GetNthItem(i);
+		printFn(std::format("{}: {} {}\n", item->idx, item->name.CStr(), usedVars.contains(item->name.CStr()) ? "" : "(unused)"));
+	}
+	printFn("\n");
+
+	// Refs
+	printFn("[Refs]\n");
+	for (int i = 0; i < script->refList.Count(); i++) {
+		const auto ref = script->refList.GetNthItem(i);
+		if (ref->varIdx) {
+			printFn(std::format("{}: (Var {})\n", i, ref->varIdx));
+		} else {
+			printFn(std::format("{}: {}\n", i, ref->form->GetEditorID()));
+		}
+	}
+	printFn("\n");
+
+	// Script data
+	printFn("[Data]\n");
+	for (int i = 0; i < script->info.dataLength; i++) {
+		printFn(std::format("{:02x} ", script->data[i]));
+	}
+	printFn("\n");
+
+	printFn(std::format("\nNum compiled bytes: {}\n", script->info.dataLength));
+
+	return true;
+}
+
+void NVSECompiler::visitNVSEScript(const NVSEScript* nvScript) {
+	// Compile the script name
+	scriptName = nvScript->name.lexeme;
+
+	// Dont allow naming script the same as another form, unless that form is the script itself
+	auto comp = strcmp(scriptName.c_str(), originalScriptName);
+	if (resolveObjectReference(scriptName, false) && comp) {
+		throw std::runtime_error(std::format("Error: Form name '{}' is already in use.\n", scriptName.c_str()));
+	}
+
+	add_u32(0x1D);
+
+	for (auto& global_var : nvScript->globalVars) {
+		global_var->accept(this);
+	}
+
+	for (auto& block : nvScript->blocks) {
+		block->accept(this);
+	}
+}
+
+void NVSECompiler::visitBeginStatement(const BeginStmt* stmt) {
+	// Get opcode
+	auto beginInfo = mpBeginInfo[stmt->name.lexeme];
+
+	// OP_BEGIN
+	add_u16(0x10);
```

Comment:
better to use an enum value than literals and comments, think there's one in [ScriptAnalyzer](https://github.com/xNVSE/NVSE/blob/master/nvse/nvse/ScriptAnalyzer.h#L55)